app/validators/case_link_type_validator.rb

```diff
@@ -0,0 +1,68 @@
+class CaseLinkTypeValidator < ActiveModel::Validator
+  ALLOWED_LINKS_BY_TYPE = {
+    related: {
+      'Case::FOI::Standard'         => ['Case::FOI::Standard',
+                                        'Case::FOI::ComplianceReview',
+                                        'Case::FOI::TimelinessReview',
+                                        'Case::ICO::FOI'],
+      'Case::FOI::TimelinessReview' => ['Case::FOI::Standard',
+                                        'Case::FOI::ComplianceReview',
+                                        'Case::FOI::TimelinessReview',
+                                        'Case::ICO::FOI'],
+      'Case::FOI::ComplianceReview' => ['Case::FOI::Standard',
+                                        'Case::FOI::ComplianceReview',
+                                        'Case::FOI::TimelinessReview',
+                                        'Case::ICO::FOI'],
+      'Case::ICO::FOI'              => ['Case::FOI::Standard',
+                                        'Case::FOI::ComplianceReview',
+                                        'Case::FOI::TimelinessReview',
+                                        'Case::ICO::FOI'],
+      'Case::SAR'                   => ['Case::SAR',
+                                        'Case::ICO::SAR'],
+      'Case::ICO::SAR'              => ['Case::SAR',
+                                        'Case::ICO::SAR'],
+    },
+    original: {
+      'Case::ICO::FOI' => ['Case::FOI::Standard'],
+      'Case::ICO::SAR' => ['Case::SAR'],
+    },
+  }.with_indifferent_access
+
+  class << self
+    def classes_can_be_linked_with_type?(type:, klass:, linked_klass:)
+      if klass.to_s.in?(ALLOWED_LINKS_BY_TYPE[type])
+        linked_klass.to_s.in?(ALLOWED_LINKS_BY_TYPE[type][klass.to_s])
+      else
+        false
+      end
```

Comment:
again, would be a bit clearer if `type` were renamed `link_type` - had me scratching my head there for a moment working out what the difference between `type` and `klass` was

Reply:
Hrm. Ya, I see why you made the earlier comment then. Is it worth renaming now?

Reply:
Let's do it as a separate thing.  Good to get this merged in ASAP I think.